django_lti_tool_provider/views.py

```diff
@@ -0,0 +1,148 @@
+
+    def get(self, request, *args, **kwargs):
+        return self.process_request(request)
+
+    def post(self, request, *args, **kwargs):
+        return self.process_request(request)
+
+    def process_request(self, request):
+        if not request.user.is_authenticated():
+            try:
+                lti_parameters = self._get_lti_parameters_from_request(request)
+                lti_data = {
+                    hook_name: lti_parameters.get(lti_name, None)
+                    for lti_name, hook_name in self.PASS_TO_AUTHENTICATION_HOOK.items()
+                }
+            except oauth2.Error, e:
+                _logger.exception(u"Invalid LTI Request")
+                return HttpResponseBadRequest("Invalid LTI Request: " + e.message)
+
+            for hook in self.authentication_hooks:
+                hook(request, lti_data)
+
+        if request.user.is_authenticated():
+            return self.process_authenticated_lti(request)
+        else:
+            return self.process_anonymous_lti(request)
+
+    @classmethod
+    def _get_lti_parameters_from_request(cls, request):
+        provider = DjangoToolProvider(settings.CONSUMER_KEY, settings.LTI_SECRET, request.POST)
+        provider.valid_request(request)
+        return provider.to_params()
+
+    @classmethod
+    def _get_anonymous_redirect_url(cls, request):
+        try:
+            result = "{login}?next={lti_handler}".format(
+                login=reverse('django.contrib.auth.views.login'),
+                lti_handler=request.path.strip('/')
+            )
+        except NoReverseMatch:
+            result = "/"
+        return result
+
+    @classmethod
+    def _get_authenticated_redirect_url(cls):
+        redirect_to = getattr(settings, 'REDIRECT_AFTER_LTI', '')
+        try:
+            result = reverse(redirect_to)
+        except NoReverseMatch:
+            result = "/"+redirect_to
+        return result
+
+    @classmethod
+    def register_authentication_hook(cls, hook):
+        """ Adds hook to pre-processing hooks list """
+        cls.authentication_hooks.append(hook)
+
+    @classmethod
+    def process_anonymous_lti(cls, request):
+        """
+        This method handles LTI request if it was sent prior to MyDante authorization. In such a case, we need user
+        authenticated first. Unfortunately, it looses POST data in the process, so when it gets back original LTI
+        request is gone. So we save important parts of it into session to retrieve when OpenID authentication happened
+        """
+        try:
+            provider = DjangoToolProvider(settings.CONSUMER_KEY, settings.LTI_SECRET, request.POST)
+            provider.valid_request(request)
+        except oauth2.Error, e:
+            _logger.exception(u"Invalid LTI Request")
+            return HttpResponseBadRequest("Invalid LTI Request: " + e.message)
+
+        request.session['lti_parameters'] = provider.to_params()
```

Comment:
@smarnach lti parameters are stored in session only during first non-authenticated request. So, the intention here is to store the parameters for later to allow user to authenticate, than redirect back to lti to handle the parameters.

So, the scenario here is as follows:

1. User clicks to open the tool from LMS for the first time.
2. It does not have cookies for the tool yet, so anonymous session is created for him, than he redirected to authentication screen.
3. Upon authentication he's redirected back to lti handler to store LTI parameters. Since this, lti parameters in session are never used (and in fact, deleted).

Te only problematic scenario I can think of here is when user opens several different assignments (or whatever) while not being authenticated. In such case there's nothing we can do.

With dalite it would be even simpler - there would be a hook to automatically authenticate user, so this code will never be executed for dalite.